src/cache/cache_las.c

```diff
@@ -1101,65 +1110,85 @@ __wt_las_sweep(WT_SESSION_IMPL *session)
 
-            /*
-             * Expect an update entry with:
-             *	1. not in a prepare locked state
-             *	2. durable timestamp as not max timestamp.
-             *	3. for an in-progress prepared update, durable
-             *	timestamp should be zero.
-             *	4. no restriction on durable timestamp value
-             *	for other updates.
-             */
-            WT_ASSERT(session, prepare_state != WT_PREPARE_LOCKED &&
-                durable_timestamp != WT_TS_MAX &&
-                (prepare_state != WT_PREPARE_INPROGRESS || durable_timestamp == 0));
+        /*
+         * Check to see if the page or key has changed this iteration, and if they have, setup
+         * context for safely removing obsolete updates.
+         */
+        remove_entry = false;
+        if (prev_entry_can_remove) {
+            if (las_btree_id != saved_btree_id || saved_key->size != las_key.size ||
+              memcmp(saved_key->data, las_key.data, las_key.size) != 0) {
+                remove_entry = true;
+                saved_btree_id = las_btree_id;
+                WT_ERR(__wt_buf_set(session, saved_key, las_key.data, las_key.size));
+            } else {
+                if (upd_type != WT_UPDATE_MODIFY)
+                    remove_entry = true;
+                else
+                    skip_update_modify_cnt++;
+            }
+        }
 
-            WT_ASSERT(session,
-              (prepare_state == WT_PREPARE_INPROGRESS || durable_timestamp >= las_timestamp));
+        /* Remove the obsolete entries */
+        if (remove_entry) {
+            /* Rewind the cursor to a point of remove record. */
+            for (i = 0; i < skip_update_modify_cnt + 1; i++)
+                cursor->prev(cursor);
+
+            WT_ERR(cursor->get_key(cursor, &las_btree_id, &las_key, &las_timestamp, &las_txnid));
+            WT_ERR(
+              cursor->get_value(cursor, &durable_timestamp, &prepare_state, &upd_type, &las_value));
+
+            /* Remove each individual LAS record till the current record. */
+            for (i = 0; i < skip_update_modify_cnt + 1; i++) {
+                __wt_verbose(session, WT_VERB_LOOKASIDE_ACTIVITY,
+                  "Sweep removing lookaside entry with "
+                  "btree ID: %" PRIu32 " saved key size: %" WT_SIZET_FMT ", record type: %" PRIu8
+                  " transaction ID: %" PRIu64,
+                  las_btree_id, saved_key->size, upd_type, las_txnid);
+
+                cursor->remove(cursor);
+                cursor->next(cursor);
+                WT_ERR(
+                  cursor->get_key(cursor, &las_btree_id, &las_key, &las_timestamp, &las_txnid));
+                WT_ERR(cursor->get_value(
+                  cursor, &durable_timestamp, &prepare_state, &upd_type, &las_value));
+                ++remove_cnt;
+            }
+            skip_update_modify_cnt = 0;
+        }
```

Comment:
This loop is going to be unnecessarily expensive in order to support `WT_VERB_LOOKASIDE_ACTIVITY` (which is never configured in the field).

I would suggest only retrieving the key  and value in the case of 
```
if (WT_VERBOSE_ISSET(session, WT_VERB_LOOKASIDE_ACTIVITY))
```
I think we could also only retrieve the key/value once, at the start of the `for` loop, rather than twice, before the for loop and at the end of the for loop?

And maybe include `remove_cnt` as part of the for loop?

Reply:
Thanks. That really improves. Fixed.